Feature/file uploader abstractions

kajarenc opened the pull request:
<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
This PR change is a reworking of a process to upload files for `st.file_uploader` and `st.camera_input`. 
It introduces the abstraction for UploadedFileManager,  the default implementation continues to store files in memory. 

Now the process of uploading file will consist from two steps
1. Issuing upload file URLs (happen via web socket communication, `file_urls_request` handler). 
2. Upload a file to the URL issued in step 1. 

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python) DONE
- E2E Tests DONE
- Any manual testing needed? DONE , testing app in Streamlit Cloud: https://file-uploader-test-k6mmsgecmepiecqfsnquhk.streamlit.app/

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

vdonato committed:
Add uuid as a dependency (#6789)

sfc-gh-kjavadyan committed:
Feature/uploaded file manager protocol (#6762)

* Uploaded file manager protocol

* Implement MemoryUploadFileManager

* Fix failing tests, not related to file uploader

* Add some tests for UploadFileRequestHandlerTest

* formatter fixes

* fix all tests in  uploaded_file_request_handler.py

* Tweak some comments and add some TODOs

* Add another TODO

---------

Co-authored-by: Karen Javadyan <kajarenc@gmail.com>
Co-authored-by: Vincent Donato <vincent@streamlit.io>

vdonato committed:
Start supporting configurable file upload URLs in StreamlitEndpoints and FileUploadClient (#6754)

* Start supporting configurable file upload URLs in StreamlitEndpoints and FileUploadClient

* Clarify some comments and fix formatting

* Remove widgetId from file upload related methods

* Have csrfRequest assume it always gets an absolute URL

vdonato committed:
Allow file uploader URLs to be fetched via websocket (#6791)

* Add new proto definitions for file upload URL requests/responses

* Add server websocket handler for file_urls_request BackMsgs

* Rename FileUrls* -> FileURLs*

* Install uuid types

* Mark resolver fields as readonly

* Allow file URLs to be requested via websocket

* Add temporary  # type: ignore

* Remove HTTP endpoints for fetching file URLs

* Remove unused variable to appease eslint

vdonato committed:
Tweak TODO

kajarenc committed:
Backend changes (#6819)

* * add `file_delete_url` to `UploadedFileInfo` proto
* add `get_upload_urls` to `UploadedFileManager`
* `UploadedFileManager` now works with file_id to get_files.

* Add todos

* remove abstractmothod from get_upload_urls

vdonato committed:
Have UploadedFileManager react client fetch and use upload URLs (#6827)

* Have UploadedFileManager react client fetch and use upload URLs

* Fix issues from rebase

vdonato committed:
Add a TODO

kajarenc committed:
Camera input with new file_uploader abstractions (#6862)

* reimplement camera_input to work with new uploaded file manager

* call delete endpoint when clear camera_input

vdonato committed:
Fix a bunch of type errors (#6885)

* Change some types from Promise<number> to Promise<void>

* Remove serverFileId and related fields from FileUploader and friends

* Remove serverFileId from CameraInput

* Remove more newly unused fields from CameraInput and FileUploader

kajarenc committed:
Feature/file uploader file urls (#6902)

* extract FileUrls to separate proto

* use copy from

* use FileURLS also in camera_input

* fixes after review

vdonato committed:
Start work on fixing file uploader and camera input e2e tests (#6891)

* Run autoformatter on e2e/specs/st_file_uploader.spec.js

* Fix some more small type errors

* Fix watched cypress route in st_file_uploader.spec.js

vdonato committed:
Fix issue with file uploader dropping every other file (#6912)

kajarenc committed:
File uploader/docs and tests (#6945)

docs and tests

vdonato committed:
Update yarn.lock

vdonato committed:
Fix import

vdonato committed:
Fix a bunch of FileUploader js unit tests (#7019)

* Fix a bunch of FileUploader js unit tests

* Fix a bunch of type errors in tests

vdonato committed:
Remove some unneeded TODOs

kajarenc committed:
fix some code ql warnings (#7031)

kajarenc committed:
add patch address

vdonato committed:
Remove outdated test

vdonato commented on lib/streamlit/runtime/runtime.py:
```diff
@@ -256,19 +258,6 @@ def get_client(self, session_id: str) -> Optional[SessionClient]:
             return None
         return session_info.client
 
-    def _on_files_updated(self, session_id: str) -> None:
-        """Event handler for UploadedFileManager.on_file_added.
-        Ensures that uploaded files from stale sessions get deleted.
-
-        Notes
-        -----
-        Threading: SAFE. May be called on any thread.
-        """
-        if not self._session_mgr.is_active_session(session_id):
-            # If an uploaded file doesn't belong to an active session,
-            # remove it so it doesn't stick around forever.
-            self._uploaded_file_mgr.remove_session_files(session_id)
```
Hm, I feel like this might be something that we want to keep even in the new world. We still have the mechanism that calls `remove_session_files` when a session is shut down, but I'm slightly worried of the possibility that a file  corresponding to a now nonexistent session is somehow uploaded so is never cleaned up without this.

Would it be possible to revive this callback function but change it to be called when no corresponding session (active or inactive) exists for a given `session_id`? This happens when `self._session_mgr.get_session_info(session_id) is None`

kajarenc replied:
Hey!
We check that session is active after the file is fully uploaded, and just before storing it into storage. 
https://github.com/streamlit/streamlit/pull/7025/files#diff-3e1711ee513b43ba25f0f9118bfe0135a503ed40da6474b7b2c5ddcef9e940f5R101 , so ~~it shouldn't be possible~~ very unlikely to upload file not corresponding to active sesion

So I don't want to make it part of the protocol but will be happy to return to this during refactoring of removing `session_id` from URLs, and rethink deleting files mechanism for open source implementation.

vdonato committed:
Fix CodeQL errors, docstrings, and other misc things

kajarenc committed:
add uuid back to dependency list

kajarenc committed:
Fix cypress failing tests, jstests and regenerate Notices (#7075)

* Wait until file uploaded instead of fix amount of time

* fix tests for CameraInput
fix tests for DefaultStreamlitEndpoints

* fix eslint errors

* add uuid to NOTICES

kajarenc committed:
Remove updated TODOs

vdonato commented on lib/tests/streamlit/elements/file_uploader_test.py:
```diff
@@ -156,40 +170,45 @@ def test_unique_uploaded_file_instance(self, get_file_recs_patch):
+        self.assertEqual(result_2, [])
+
+    @patch("streamlit.elements.widgets.file_uploader._get_upload_files")
+    def test_deleted_files_filtered_out(self, get_upload_files_patch):
```
Aren't these two tests essentially identical? I think we can probably remove the first one

kajarenc replied:
Good catch, the second test is a superset of the first test, I removed the first one!

vdonato committed:
Fix docstring

kajarenc committed:
fix docstring and remove unused import

kajarenc commented on lib/streamlit/runtime/state/session_state.py:
```diff
@@ -179,6 +179,8 @@ def set_widget_metadata(self, widget_meta: WidgetMetadata[Any]) -> None:
     def remove_stale_widgets(self, active_widget_ids: set[str]) -> None:
         """Remove widget state for widgets whose ids aren't in `active_widget_ids`."""
+        # TODO(vdonato / kajarenc): Remove files corresponding to an inactive file
+        # uploader.
```
Not critical, we keep this for future improvement

vdonato commented on lib/streamlit/elements/widgets/camera_input.py:
```diff
@@ -33,18 +33,18 @@
-def _get_file_recs_for_camera_input_widget(
-    widget_id: str, widget_value: Optional[FileUploaderStateProto]
-) -> List[UploadedFileRec]:
+def _get_upload_files(
```
Can we factor this out into a helper function shared by `camera_input` and `file_uploader`? This function seems to be identical for both of the widgets

kajarenc replied:
Yes, it difficult to find a new home for `def _get_upload_files` (because it is used only in file_uploader and camera_input, so `utils.py` is probably not the best place, so I just now imported `_get_upload_files` from file_uploader for camera_input,  I think it is good idea, and denote the fact that camra_input essentially a derivative form file_uploader

vdonato commented on lib/streamlit/elements/widgets/file_uploader.py:
```diff
@@ -37,11 +36,12 @@
 
-SomeUploadedFiles = Optional[Union[UploadedFile, List[UploadedFile]]]
-
+SomeUploadedFiles = Optional[
+    Union[UploadedFile, DeletedFile, List[Union[UploadedFile, DeletedFile]]]
+]
```
nit: can add `None` to the union instead of having an `Optional[Union[...]]` to be consistent with camera input

kajarenc replied:
done

vdonato commented on lib/streamlit/elements/widgets/file_uploader.py:
```diff
@@ -442,21 +446,20 @@ def _file_uploader(
-
         self.dg._enqueue("file_uploader", file_uploader_proto)
-        return widget_state.value
+
+        filtered_value: Union[UploadedFile, List[UploadedFile], None]
+
+        if isinstance(widget_state.value, DeletedFile):
+            filtered_value = None
+        elif isinstance(widget_state.value, list):
+            filtered_value = [
+                f for f in widget_state.value if not isinstance(f, DeletedFile)
+            ]
+        else:
+            filtered_value = widget_state.value
+
+        return filtered_value
```
nit: I think it'd be fine to return the value directly from each branch of the if/elif/else statement rather than save it to an intermediate variable that's returned immediately

kajarenc replied:
done

vdonato commented on lib/streamlit/elements/widgets/file_uploader.py:
```diff
@@ -225,7 +229,7 @@ def file_uploader(
         disabled: bool = False,
         label_visibility: LabelVisibility = "visible",
-    ) -> SomeUploadedFiles:
+    ) -> Optional[Union[UploadedFile, List[UploadedFile]]]:
```
nit: same comment about adding `None` to the union instead of having this be `Optional` (same thing again with L389)

kajarenc replied:
done

vdonato commented on lib/streamlit/runtime/uploaded_file_manager.py:
```diff
@@ -11,324 +11,118 @@
+class DeletedFile(NamedTuple):
+    """Represents a deleted file in deserialized values for st.file_uploader and
+    st.camera_input
+    Return this from st.file_uploader and st.camera_input deserialize (so they can
```
nit: should have a newline between the docstring summary and the rest of it

kajarenc replied:
done

vdonato reviewed (commented):
Some small comments, but overall this LGTM

I won't officially hit the +1 button given so much of this code is mine that I'm pretty sure that I'm not qualified to approve it 😆

kajarenc committed:
Merge branch 'remote-develop' into feature/file_uploader_abstractions

# Conflicts:
#	lib/streamlit/elements/widgets/camera_input.py
#	lib/streamlit/elements/widgets/file_uploader.py

kajarenc committed:
Changes after review

kajarenc committed:
fix docstrings

kajarenc committed:
replace Optional[Union[X]], with Union[X, None]

lukasmasuch commented on frontend/lib/src/FileUploadClient.ts:
```diff
@@ -31,10 +36,11 @@ interface Props {
 
 /**
- * Handles uploading files to the server.
+ * Handles operations related to the st.file_uploader widget.
```
Isn't it also used for `st.camera_input`? If so, maybe rewrite as something like:

```suggestion
 * Handles operations related to the widgets that require file uploading.
```

lukasmasuch commented on frontend/lib/src/StreamlitEndpoints.ts:
```diff
@@ -34,6 +34,14 @@ export interface StreamlitEndpoints {
 
+  /**
+   * Construct a URL for uploading a file.
+   * @param url a relative or absolute URL. If `url` is absolute, it will be
```
nit: I think in the [JSdoc docstring syntax](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns) the parameter is separated via a `-`, e.g.:

```suggestion

   * @param url - a relative or absolute URL. If `url` is absolute, it will be
```

There are also a few other instances in the PR that could get correct, but it seems that we haven't been following this anyways in the current impl.

lukasmasuch commented on proto/streamlit/proto/Common.proto:
```diff
@@ -46,17 +46,48 @@ message StringTriggerValue {
 
 message FileUploaderState {
+  // DEPRECATED
   sint64 max_file_id = 1;
```
Same as the comment above, couldn't we just fully remove this via:

```suggestion
  reserved 1;
  reserved "max_file_id";
```

kajarenc replied:
I am not sure we are okay to remove / change the type of any proto message field, because of dependencies from other teams.

This is a good question to clarify, to have a solid understanding of the contract we keep for proto messages. 
For now, I think the safest thing is just to keep the field as is, with a comment that it is now deprecated.

lukasmasuch replied:
oki, let's keep it as is 👍 But maybe we bring it up in a standup soon to figure out if it would break other dependencies.

lukasmasuch commented on proto/streamlit/proto/Common.proto:
```diff
@@ -46,17 +46,48 @@ message StringTriggerValue {
 // Information on a file uploaded via the file_uploader widget.
 message UploadedFileInfo {
+  // DEPRECATED.
   sint64 id = 1;
```
Is this field still used in any way? If not, could we just remove this and add:

```suggestion
  reserved 1;
  reserved "id";
```

lukasmasuch commented on lib/streamlit/elements/widgets/file_uploader.py:
```diff
@@ -137,7 +144,7 @@ class FileUploaderMixin:
         self,
         label: str,
-        type: Optional[Union[str, Sequence[str]]],
+        type: Union[str, Sequence[str], None],
```
nit: we could update this to the new annotation format:

```suggestion
        type: str | Sequence[str] | None,
```

(requires `from __future__ import annotations`)

kajarenc replied:
Yes, this is something I considered, but at the top of this file we have 

```python
SomeUploadedFiles = Union[
    UploadedFile,
    DeletedFile,
    List[Union[UploadedFile, DeletedFile]],
    None,
]
```

and even with `from __future__ import annotations` 
```python
SomeUploadedFiles = UploadedFile | DeletedFile | List[Union[UploadedFile, DeletedFile]] | None
```
doesn't work, because we have assignments and do not just type annotation

To keep it consistent, I do not import annotations for that files, although I don't think it is a very bad idea to not be inconsistent in this case

lukasmasuch commented on lib/streamlit/runtime/memory_uploaded_file_manager.py:
```diff
@@ -0,0 +1,135 @@
+        all_files: List[UploadedFileRec] = []
+        # Make copy of self.file_storage for thread safety, to be sure
+        # that main storage won't be changed form other thread
+        file_storage_copy = self.file_storage.copy()
```
nit: does that mean that all files are also duplicated in memory? So calling this (e.g. via the metrics endpoint), might crash the app because of a sudden memory increase. Not a big deal right now since it isn't used a lot, but if we do more with those stats we might want to revisit this.

kajarenc replied:
yes, now we create a duplicate to collect statistics. 
The good news is that this duplicate should be short-living, and Python Garbage Collector should collect it very fast because we don't keep any references to `file_storage_copy`, but yes, this could lead to memory usage increase in case of a lot of stored uploaded files.

lukasmasuch replied:
Maybe you can add a TODO comment or something there, so that we double check this once we do more with the stats.

lukasmasuch commented on lib/streamlit/runtime/uploaded_file_manager.py:
```diff
@@ -11,324 +11,119 @@
+        """Remove all files associated with a given session."""
+        raise NotImplementedError
+
+    def get_upload_urls(
```
nit: maye also add full docstring for the params and return value, since this is kind of an official interface that might also be implemented by other teams.

kajarenc replied:
docstring added!

lukasmasuch commented on frontend/app/src/App.tsx:
```diff
@@ -1525,6 +1529,20 @@ export class App extends PureComponent<Props, State> {
 
+  requestFileURLs = (requestId: string, files: File[]): void => {
+    if (this.isServerConnected()) {
+      const backMsg = new BackMsg({
```
This back message doesn't trigger a rerun, or?

kajarenc replied:
correct, it doesn't trigger a rerun

lukasmasuch commented on frontend/app/src/App.tsx:
```diff
@@ -1525,6 +1529,20 @@ export class App extends PureComponent<Props, State> {
+      const backMsg = new BackMsg({
+        fileUrlsRequest: {
+          requestId,
+          fileNames: files.map(f => f.name),
```
What happens if a user has two file upload widgets and uploads a file with the same name to both widgets? Does this work with the new impl. as well?

kajarenc replied:
Yes, it works correctly, because for each file we issue on the backend a different `file_id` as part of file URLs, so for the next steps from the streamlit perspective they an absolutely different files.

lukasmasuch reviewed (approved):
LGTM 👍  I just have some nits and questions, but no major concern.

kajarenc committed:
add docstring

kajarenc committed:
Update frontend/lib/src/FileUploadClient.ts

Co-authored-by: Lukas Masuch <Lukas.Masuch@gmail.com>

vdonato committed:
Don't require new methods/args in StreamlitEndpoints and FileUploadClient (#7092)

We want to avoid having the notebooks team have to make any code changes on their end due to the
interface / constructor changes we're making to StreamlitEndpoints and FileUploadClient. In order
to do this, we make the newly added methods/args optional.
